tools/ThermoRawFileParser/thermo_converter.xml

```diff
@@ -1,60 +1,227 @@
+            <filter>output_format == "1"</filter>
+            <filter>len(input) > 1</filter>
+            <filter>gzip_boolean == 0</filter>
+            <discover_datasets pattern="(?P&lt;designation&gt;.+)\.mzML" ext="mzml" directory="output_folder"/>
+        </collection>
+
+        <collection name="output_indexedmzml_collection" type="list" label="${tool.name} on ${on_string}: Indexed mzML">
+            <filter>output_format == "2"</filter>
+            <filter>len(input) > 1</filter>
+            <filter>gzip_boolean == 0</filter>
+            <discover_datasets pattern="(?P&lt;designation&gt;.+)\.mzML" ext="mzml" directory="output_folder"/>
+        </collection>
+        <!--
+        <collection name="output_parquet_collection" type="list" label="${tool.name} on ${on_string}: parquet">
+            <filter>output_format == "3"</filter>
+            <filter>len(input) > 1</filter>
+            <discover_datasets pattern="(?P&lt;designation&gt;.+)\.parquet" directory="output_folder"/>
+        </collection>
+      -->
+        <collection name="output_gzipped_collection" type="list" label="${tool.name} on ${on_string}: Gzipped data">
+            <filter>len(input) > 1</filter>
+            <filter>gzip_boolean == 1</filter>
+            <discover_datasets pattern="(?P&lt;designation&gt;.+)\.gz" ext="zip" directory="output_folder"/>
+            <discover_datasets pattern="(?P&lt;designation&gt;.+)\.gzip" ext="zip" directory="output_folder"/>
+        </collection>
+
+        <collection name="output_metadata_collection" type="list" label="${tool.name} on ${on_string}: metadata">
+            <filter>output_metadata_selector != "off"</filter>
+            <filter>len(input) > 1</filter>
+            <discover_datasets pattern="(?P&lt;designation&gt;.+)-metadata.txt" ext="txt" directory="output_folder"/>
+            <discover_datasets pattern="(?P&lt;designation&gt;.+)-metadata.json" ext="json" directory="output_folder"/>
+        </collection>
+
     </outputs>
+
     <tests>
+        <!-- Basic test -->
         <test>
-            <param name="input" value="fake_input.txt"/>
-            <assert_command>
-                <has_text text="ThermoRawFileParser.sh"/>
-            </assert_command>
+            <param name="input" value="really_small.raw"/>
+            <param name="output_format" value="1"/>
+            <output name="output" file="really_small.mzml" ftype="mzml" compare="sim_size" delta="3000"  >
+            </output>
         </test>
+
+        <!-- Basic gzipping test -->
+        <test>
+            <param name="input" value="really_small.raw"/>
+            <param name="output_format" value="1"/>
+            <param name="gzip_boolean" value="1"/>
+            <output name="output_gzipped" file="really_small.mzml.gzip" ftype="zip" compare="sim_size" delta="3000"  >
+            </output>
+        </test>
+
+        <!-- Testing contents of converted mgf file-->
         <test>
```

Comment:
you can use `expect_num_outputs` to enhance your tests for all your output filers.

Reply:
Um... using that option, do you know if there is a way to know which outputs are found when running the tests? I am surprisingly having errors on all tests (on both non zipped and zipped output tests) when using "expect_num_outputs="1":
`
Exception: Incorrect number of outputs - expected 1, found 2.
`

I do not know how it may be possible...

Reply:
I recommend grepping in https://github.com/galaxyproject/galaxy/tree/dev/test/functional/tools and getting inspiration. This should work. If you have collection the elements count differently. I think it was every element + 1 for the collection.

Reply:
I am quite surprised with the way it is working.

This works:
```

        <!-- Basic mzml collection test -->
        <test expect_num_outputs="1">
            <param name="input" value="really_small.raw,really_small_2.raw"/>
            <param name="output_format" value="1"/>
            <output_collection name="output_mzml_collection" type="list" count="2"/>
        </test>
        
        <!-- mgf collection test with metadata -->
        <test expect_num_outputs="2">
            <param name="input" value="really_small.raw,really_small_2.raw,really_small_3.raw"/>
            <param name="output_format" value="0"/>
            <param name="output_metadata_selector" value="0"/>
            <output_collection name="output_mgf_collection" type="list" count="3"/>
            <output_collection name="output_metadata_collection" type="list" count="3"/>
        </test>
```

So, at least with collections, it seems that "expect_num_outputs" just count the number of <output_collection> elements with content in the output.

But both single input tests, as
```

<!-- Basic test -->
        <test expect_num_outputs="1">
            <param name="input" value="really_small.raw"/>
            <param name="output_format" value="1"/>
            <output name="output" file="really_small.mzml" ftype="mzml" compare="sim_size" delta="3000" />
        </test>
```
fails with

Incorrect number of outputs - expected 1, found 2.

I don’t see any way of that test producing 2 outputs… apart from being also counting the “output_metadata” output (although it is empty)?

Reply:
> I recommend grepping in https://github.com/galaxyproject/galaxy/tree/dev/test/functional/tools and getting inspiration. This should work. If you have collection the elements count differently. I think it was every element + 1 for the collection.

I am doing it but it do not find many uses of "expect_num_outputs" specially mixed with single outputs, collections and things like "change_format"...

Reply:
Try this one: https://github.com/galaxyproject/tools-iuc/search?l=XML&q=expect_num_outputs

Reply:
Mh, does planemo server generate the metadata output? Maybe <filter>output_metadata_selector != "off"</filter> is wrong and it needs to be `<filter>str(output_metadata_selector) != "off"</filter>`

Reply:
Tank you for the examples... which metadata do you mean?
I am doing many checks and I have think that filter "len(input)>1" is not working/filtering when doing tests (but it seems to work perfectly when executing the tool) as output_metadata_collection seems to be always present even when I introduced just one input...
If I put some impossible condition in output_metadata_collection like <filter>"1" == "2"</filter> then output_metadata_collection is not generated and the number of outputs when generating single "output" and "output_metadata"  is 2.. If I do not specify that kind of impossible condition, expect_num_outputs is 3.

Reply:
Well, after some try & error, it seems the filter works wih
`
(str(input)).count(',') == 0
`
:-)